7.x 1.2 beta

AntonioXIII opened the pull request:
This branch addresses the following:

- BMM block HTML is no longer generated in three separate places (which creates a maintenance nightmare). The block HTML is now generated with one function in a new helpers.php file.
- Menu items can now be pointed at <front> and have a dropdown.
- Multiple menu items can now be pointed at the same link, so long as they don't have the same title.

@skenley Please verify through your own testing before merging. I have already undergone my own tests of the new version, and everything passed.

AntonioXIII committed:
Condense Block HTML Build Code

Before this commit, the block HTML being generated for the BMM was being
repeated in three different places. A new file has been added in an
includes file called helpers, which contains the helper function that is
now solely responsible for block HTML building.

AntonioXIII committed:
Item Pointed at Home Dropdown Fix

When a menu link at the 1st level pointed to home and had dropdown
items, the dropdown wasn't displayed. This was because of an issue in
the better_mm_get_children function, which called
menu_link_get_preferred method to fetch the path of the home link for
building the menu tree. However, the function returned a non-existant
item from the navigation menu, which resulted in no dropdown. This was
fixed by using menu_tree_all_data to fetch the menu structure manually
from the database, and return the children.

AntonioXIII committed:
Get Menu Children Changed

The issues #20 and #21 stem from the use of the code that retrieves menu
item children (also amended in the last commit). The code only checked
for a menu item's path, and then got the menu item's children using
menu_link_get_preferred. I have entirely restructured this code which
now fixes these issues. However, a break could potentially happen if
there are two menu items with the exact same title, and same path. This
should not be done though, so it is something that I am deeming
acceptable, but can be changed later if that functionality is needed for
some reason.

skenley reviewed (changes requested):
bmm_menu_items_settings, seems to be only updating with the save of the BMM settings form. Need to make sure it's updated on save of menu item form, and probably the node save.

If a menu item has the same title AND path as another item the children of the first appearing item is given to the second appearing item.

AntonioXIII commented:
Steve tested on: https://dashboard.pantheon.io/sites/bdf5c670-2a08-4b02-86a4-85691e481772#bmm-test

AntonioXIII committed:
Update BMM Form Settings

In an issue that came up during testing for pull request #24 it was
found that the BMM form settings database variable was not being updated
as menu items were deleted and added. I have added code (including a new
helper function) that will do this now.

AntonioXIII committed:
Duplicate Links and Titles

From another issue brought up in testing on pull request #24 a user
could not add two menu items with the exact same title and path. The
code has been changed now to check for mlid (menu link id) for menu item
children instead of menu link and title. The result is now that you can
have two (or more) menu items with the exact same name, and path while
maintaining each link's unique dropdown of children.

AntonioXIII committed:
Version Change and File Rename

Changed the version number in the .info file to reflect 1.2-beta and I
renamed the extension of the helpers file to be .inc instead since that
seems to be drupal standard for module include files.

AntonioXIII commented:
@skenley I have fixed the issues you found in your testing. This is now ready for testing again.

AntonioXIII committed:
Fixed Disabled Items Printing

During testing I found that disabled menu items were getting printed in
the BMM. Disabled menu items are now taken into account in the newly
refactored code for 1.2-beta.

AntonioXIII committed:
Fix Error On Wide Setting for Home Link

When uploading an image for a primary link that points to <front> an error occurs because the menu_link_path function doesn't find a link with that path. With the new design the mlid is already available, so that is used instead of retrieving it manually from the database.
